### PR TITLE
[Merged by Bors] - TY-3018-fk-violation-error-helper

### PR DESCRIPTION
**Pull request by rustonaut**
Adds two helper:

- `From<sqlx::Error>` for `storage::Error` so that we don't have to write `.await.map(|err| Error::Database(err.into()))?` all the time
- a `.fk_violation_is_invalid_document()` extension on `Result<V, sqlx::Errror>` which will convert the (potential) error to a `storage::Error` like using `into()` but if there is a fk violation error it will produce an `Error::InvalidDocumentId` variant instead of an `Error::Database`.

### CODE REVIEW

File: discovery_engine_core/core/src/storage/sqlite.rs [hunk 1]
```diff
@@ -49,6 +49,29 @@ use crate::{
 // Sqlite bind limit
 const BIND_LIMIT: usize = 32766;
 
+trait SqlxSqliteErrorExt<V> {
+    /// Use this on the result of an sqlite query which might have failed a foreign
+    /// key constraint if a illegal document if was passed in.
```
Comment (Robert-Steiner) [inline discovery_engine_core/core/src/storage/sqlite.rs:54]:
```suggestion
    /// Use this on the result of a sqlite query which might have failed a foreign
    /// key constraint when an illegal document was passed in.
```

File: discovery_engine_core/core/src/storage/sqlite.rs [hunk 2]
```diff
@@ -785,4 +808,39 @@ mod tests {
...
+        let res = sqlx::query("INSERT INTO Bar(x) VALUES (?);")
+            .bind(10)
+            .execute(&storage.pool)
+            .await
+            .fk_violation_is_invalid_document_id(document_id);
+
+        assert!(matches!(res, Err(Error::InvalidDocumentId(_))));
+        if let Err(Error::InvalidDocumentId(id)) = res {
+            assert_eq!(id, document_id);
+        }
```
Comment (Robert-Steiner) [inline discovery_engine_core/core/src/storage/sqlite.rs:837]: 🎨 
```suggestion
        assert!(matches!(res, Err(Error::InvalidDocumentId(id)) if id == document_id));
```

**Review by Robert-Steiner — APPROVED** [review]
lgtm

**Comment by rustonaut** [comment]
bors merge

**Pull request closed without merging**